Refactor the cascading code in the typed API into a self contained backend

johnynek opened the pull request:
As expected, this was actually easier than we might have suspected for a while.

This is not ready to ship. We will need to test it a fair amount. Also we need to probably add optimizations (the graph optimizer from summingbird should be great here).

But the code is much cleaner.

It is not totally factored, as I didn't move all the Grouped and CoGrouped code into the CascadingBackend, but that is easy, I just wanted to keep the diff size down a bit.

With this code, we could almost plan to any backend (when we are well factored) and just error if someone starts with a raw cascading pipe.  The backend would need a map of `TypedSource[T]` to the platform's source type. We will also need to update `Execution[_]` to deal with pluggable backends, but that should be in a separate PR in my view.

cc @piyushnarang @ianoc @isnotinvain @fwbrasil @dieu 

todo:
- [x] ~benchmark the approach we have to function composition vs the previous stack consuming approach (stacks should not be deep, so just do the absolute fastest)~ just avoid the fancy stack avoiding stuff and do it like we used to.
- [x] fix the fact that we are regressing on some line numbers currently (previously TypedPipeFactory put a line number on almost everything). We keep line numbers on grouping and writing, which is probably enough for many cases, but we should not regress (although the tests are clearly weak on this since we are only failing 1 and only on 2.11.8).

johnynek committed:
Refactor the cascading code in the typed API into a self contained backend

johnynek commented:
Also, this still blends some optimizations in the types (join composition is still handled by CoGrouped). Ideally we want to separate and do no optimizations as we build the AST, and then when we plan apply the general rules (those that apply to all graphs, e.g. `(a ++ b).map(f) == (a.map(f) ++ b.map(f))` vs those that only apply to Cascading (a GroupBy node can accept any number of input pipes).

johnynek committed:
Fix some of the test breaks

johnynek commented:
This still does not pass all the tests.

The two issues are:

1. line numbers in many cases were lost since they were handled as a side effect of how TypedPipeFactory worked.
2. optimizations around zip of Execution seems a bit fragile, so we seem to have blocked some of them here. I think it has to do with some of the pipes looking different, and I am not sure the test is exactly a real failure in that I fear there may have been other fragile cases in the past. We should fix the issue and maybe make it more robust.

In a follow up PR, we should move more of the Execution related code into `CascadingBackend`.

I'm fairly optimistic this approach could allow us to:
a. both make the planning more clear and understandable.
b. enable new, whole-graph optimizations which we can't do well now.
c. allow us to run on other platforms (spark, flink, concurrent local) with minimal changes (error if the user converts directly from a cascading pipe but plans to another platform, so we can keep source compatibility with the existing API).

johnynek commented:
another optimization that occurs to me is that we could now pretty easily avoid lifting every map function to `A => TraversableOnce[B]`. Instead if we have a `A => B` we could avoid the allocation of `Iterator.single`. Since we do that on each row, that could actually be a measurable improvement for such jobs.

johnynek committed:
fix overspecific repl test

johnynek committed:
Add a hack to fix a failing test

piyushnarang commented:
This is pretty cool, thanks for putting this up. Will take a look in the next day / two.

fwbrasil commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+    val pipe =
+      loop(p, FlatMappedFn.identity[U])
+        .applyFlowConfigProperties(flowDef)
+    RichPipe.setPipeDescriptionFrom(pipe, LineNumber.tryNonScaldingCaller)
```
would it be possible to move the line numbers to the `TypedPipe` AST instances to keep the information precise?

johnynek replied:
I think we should. I would like to add another AST node to annotate with line numbers rather than add it to each node. I wanted to keep this diff as small as possible though so people can read it.

fwbrasil replied:
👍

fwbrasil commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+  /**
+   * we want to cache renderings of some TypedPipe to Pipe so cascading
+   * will see them as the same. Without this, it is very easy to have
+   * a lot of recomputation. Ideally we would plan an entire graph
```
I'm not sure I understand the flow here. Isn't `TypedPipe.toPipe` called only once given a job?

johnynek replied:
No. It is called on each `.write` and some jobs write many outputs. It is probably also called somewhere inside Twitter by someone doing some mixing of the fields and typed API.

fwbrasil replied:
I see, thanks for the clarification.

fwbrasil commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+   * at once, and not need a static cache here, but currently we still
+   * plan one TypedPipe at a time.
+   */
+  private[this] val pipeCache = new WeakHashMap[TypedPipe[Any], Map[Mode, CascadingPipe[Any]]]()
```
a global static cache for mutable values is pretty scary :) Is there a way that we could reduce its scope? For instance, if the flow def creation calls `TypedPipe.toPipe` multiple times but using the same thread, we could use a `ThreadLocal`-bound cache that can be cleaned after the planning.

johnynek replied:
what values are mutable here?

fwbrasil replied:
aren't the cascading values mutable? (fields, flowDef)

johnynek replied:
while they are technically mutable, we are not mutating any of them as we store them in here. This is kind of the whole game with this API. We are putting immutable fascades over mutable types, but we are being sure to make sure they are not mutated.

Note: all the items added to the cache we create in this file, and we never return them back to the user, so I think it is totally safe, but the type system it not helping us prove that.

fwbrasil replied:
Cool! I thought they'd be mutated by the cascading calls. Maybe we could encapsulate the mutable values and expose only safe methods? I worry if someone mutates the values by mistake since they are accessible as `CascadingPipe` fields.

johnynek replied:
it would be interesting to think about that, but it is not totally simple. Since we have to get them back out periodically to pass to cascading. We could copy at that point since the performance of planning is not really critical (it is very fast), the runtime is totally dominated by job running time and a little bit of IO at scheduling time.

I'd say this is something to add to the roadmap for the next version (if we adopt this PR), but not a blocker for this PR, if you ask me.

fwbrasil commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+    pipeCache.synchronized {
+      pipeCache.get(t) match {
+        case null => add(Map.empty)
+        case somemap if somemap.contains(m) => somemap(m).asInstanceOf[CascadingPipe[T]]
```
It seems that some of the AST nodes have values that use the identity hash code (e.g. `CascadingPipe.flowDef`), so even though nodes may represent the same computation they may not match the same hashmap entry. I'm not sure that's a problem, but I thought it's worth pointing out.

johnynek replied:
I believe that in fact we will never put a `CascadingPipe` directly into this cache. We could assert that to make it clear. It could come in as a grouping from someone that uses `from(pipe: Pipe, f: Field)...`. That said, in that case, I think we want identity hashCode and equals. I think we can add a discussion of this in the comment.

fwbrasil replied:
I'm probably missing the big picture, but it seems fragile to rely on the identity hash code of values to determine that two AST nodes represent the same computation. Maybe this discussion doesn't make sense, though. What's the impact if eventually a hash code mismatch happens? Is it just performance of the flow planning?

johnynek replied:
if we have to semantically equal but not identical flowDefs it means we may wind up recomputing something (if cascading does not dedup at its level). We could implement a `FlowDef` wrapper that implements a hashCode and equality, but I don't think it is high priority actually.

fwbrasil commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+   * at once, and not need a static cache here, but currently we still
+   * plan one TypedPipe at a time.
+   */
+  private[this] val pipeCache = new WeakHashMap[TypedPipe[Any], Map[Mode, CascadingPipe[Any]]]()
```
Wdyt about moving the caching logic to a separate class?

johnynek replied:
I kind of want to keep it as local as possible and hopefully kill it eventually. If you prefer, I can make a class that encapsulates this WeakHashMap and lookup.

Note this previously existed in some form (and was needed to get tests to pass) in the previous version as well (inside TypedPipeFactory)

fwbrasil replied:
Fair enough, I agree that it's better to keep it limited to this scope 👍

piyushnarang replied:
Actually I was thinking a inner class would be nice too. I don't mind keeping it in this class (rather than breaking to a separate file)

fwbrasil commented on scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala:
```diff
@@ -119,6 +113,53 @@ object TypedPipe extends Serializable {
+    converter: TupleConverter[T]) extends TypedPipe[T]
+
+   case class CrossPipe[T, U](left: TypedPipe[T], right: TypedPipe[U]) extends TypedPipe[(T, U)] {
+     def viaHashJoin: TypedPipe[(T, U)] =
```
This transformation seems much simpler than the previous implementations of `cross`. Is it a safe change? Would it be worth keeping the old behavior to reduce the scope of the change?

johnynek replied:
you mean here:
https://github.com/twitter/scalding/pull/1666/files/a76865d69948a2fef0953b8a061a95da9c070bf0#diff-fbe95168deecb5a4b5e9ace30182593eL981

I'm happy to add more optimization rules here, but I think all of those (other than the Merged change) are small potatoes. doing a cross is super rare anyway (because it multiplies the number of outputs) and the optimization for using iterables via closure capture is was not motivated by benchmarks.

The real reason I kept it simple is before we launch this we will want to actually run a lot of more principled and tested optimization rules rather than ad-hoc ones and the ones on cross seemed both ad-hoc (there was even a TODO to remove them) and also rarely exercised (since cross is so rare).

fwbrasil replied:
sounds good, thanks for the clarification

fwbrasil commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+
+      case DebugPipe(p) =>
+        // There is really little that can be done here but println
+        loop(p.map { t => println(t); t }, rest)
```
How often is `.debug` used? Is it useful? Reading the code, it seems that the new implementation will output way less information.

johnynek replied:
can you explain? The old implementation just called cascading Debug pipe on things that were already scheduled, which does this:
https://github.com/Cascading/cascading/blob/3.1/cascading-core/src/main/java/cascading/operation/Debug.java#L234

which, since we have it print every item, I think is just different in that it prints it with cascading fields, which are meaningless to the user of scalding's typed API (and actually confusing).

fwbrasil replied:
I see. I saw https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/PipeDebug.scala and https://github.com/twitter/scalding/pull/559/files, and thought that we could be losing relevant debug information.

fwbrasil reviewed (commented):
Initial review. I'm still wrapping my head around the code and changes, but it looks promising! I wonder if it'd be better to reduce the change set even further and avoid behavior changes.

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/typed/FlatMappedFn.scala:
```diff
@@ -20,38 +20,73 @@ import java.io.Serializable
-case class Converter[R](conv: TupleConverter[R]) extends FlatMapFn[R] {
-  // make sure not to start with an Iterator to keep everything lazy
-  def apply(te: TupleEntry) = Iterator(conv(te))
+  case class Identity[A, B](ev: A =:= B) extends FlatMapping[A, B]
```
should we move these up? or you wanted to keep them within this object?

johnynek replied:
I like keeping subclasses of a sealed trait in the companion object to avoid polluting the outer namespace with things that should be implementation details.

piyushnarang replied:
ok, sounds good.

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/typed/FlatMappedFn.scala:
```diff
@@ -20,38 +20,73 @@ import java.io.Serializable
-case object Empty extends FlatMapFn[Nothing] {
-  def apply(te: TupleEntry) = Iterator.empty
+/**
+ * This is a composition of one or more FlattMappings
```
s/FlattMappings/FlatMappings ?

johnynek replied:
👍

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+      }
+
+    def loop[T](t: TypedPipe[T], rest: FlatMappedFn[T, U]): Pipe = t match {
+      case cp@CascadingPipe(_, fields, localFlowDef, m, _) =>
```
I'm wondering if we could split this up. This method is a super big case match which is a bit hard to read. Would it make sense to move the loop method to each of these specific subclasses? So we'd have overrides for each type of pipe?

johnynek replied:
I would really rather not do that since that puts the Pipe on the subclasses of TypedPipe which we are trying to avoid.

The case match is big, but each is a pretty clear block and we have a check that we have a total match.

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+import scala.collection.mutable.{ Map => MMap }
+import java.util.WeakHashMap
+
+object CascadingBackend {
```
should we move this to a separate package (`cascading`?)

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala:
```diff
@@ -0,0 +1,268 @@
+        IterableSource(Iterable.empty, fieldNames)(setter, singleConverter[U]).read(flowDef, mode)
+
+      case fk@FilterKeys(_, _) =>
+        def go[K, V](node: FilterKeys[K, V]): Pipe = node match {
```
maybe give this method a nicer name? doFilterKeys? / filterKeys?

piyushnarang replied:
same for the others below?

johnynek replied:
I don't want to do that because I think it actually obscures things. If scala were smarter with type inference this would not even be needed (note we just define a method to call it). I think keeping a standard pattern here actually serves to tell the reader this is doing the logic for the match, except we can name the types.

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/typed/HashJoinable.scala:
```diff
@@ -45,20 +45,23 @@ trait HashJoinable[K, +V] extends CoGroupable[K, V] with KeyedPipe[K] {
-        getForceToDiskPipeIfNecessary(fd, mode),
-        Field.singleOrdered("key1")(keyOrdering),
-        WrappedJoiner(new HashJoiner(joinFunction, joiner)))
+  //def hashCogroupOn[V1, R](mapside: TypedPipe[(K, V1)])(joiner: (K, V1, Iterable[V]) => Iterator[R]): TypedPipe[(K, R)] =
```
delete?

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/typed/HashJoinable.scala:
```diff
@@ -109,11 +112,11 @@ trait HashJoinable[K, +V] extends CoGroupable[K, V] with KeyedPipe[K] {
-          case _: MapFn[_, _] => false
-          case _: FlatMappedFn[_, _] => false
-          case _ => false // treat empty fn as a Filter all so forceToDisk
+          case fmp: FlatMappedFn[_, _] if (FlatMappedFn.asId(fmp).isDefined) =>
```
let's also update the comment at the top of this method. It refers to the old converter / emptyFn etc.

johnynek committed:
Address review comments, fix line number issue

johnynek commented:
okay, other than benchmarking the new function composition, I think this addresses most comments.

I think after the benchmark, this should be mergeable. Can you look again @fwbrasil and @piyushnarang ?

johnynek committed:
Don't use an interpretter in function composition

johnynek commented:
turns out of benchmarks don't compile, which is a bummer. Rather than shave that yak right now, I'm just making the function composition use `andThen` to compose except be more efficient than before (avoid allocating a Iterator except at the very end for map sequences, be smarter about filter as well).

I think this is mergeable. What do you folks say?

fwbrasil commented:
LGTM! The build is failing, though.

johnynek committed:
Fix an overspecific test

piyushnarang commented:
👍 This looks good to me as well.
